### ISSUE TITLE
Generate an error when my function crash

### ISSUE THREAD

**Issue by nico-robert** [apnadkarni/tcl-cffi#64]
Hello @apnadkarni ,

I am not familiar with the C language and probably also my misunderstanding of your documentation.
The library I want to use, generate an error when a function does not work with as message : 
`ERROR: error_no=**** detail_no=*`

The definition of an error handler is as follows :
```
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <setjmp.h>

jmp_buf env;

error_handler  (_STATUS   error_no,
                _STATUS   detail_no,
                void      *user_data)
{
    printf ("ERROR: error_no=%04X, detail_no=%u\n", (_UINT)error_no,
                (_UINT)detail_no);
    longjmp(env, 1);
}
```
I would like to reproduce the same thing... I tried to apply this using your following example _examples/libzip.tcl_
without success.

My C function with unknown path  : 
`image = _LoadPngImageFromFile (my_pointer, "*****/p02.png");`
return :  _ERROR: error_no=1017, detail_no=2_

My tcl function : 
```
cffi::Struct create error_handler {
    error_no  ulong
    detail_no ulong
    user_data {pointer unsafe}
}
cffi::alias define _ErrorHandler {struct.error_handler byref}
LIB stdcall _LoadPngImageFromFile pointer.int {p pointer.int filemane string}
# Tcl function :
set image [_LoadPngImageFromFile $my_pointer "*****/p02.png"]
```
return : _Bus error: 10_ (tclsh is closed with this message (I'am on MacOS))

I already use this structure when initializing my first pointer and I feel it works (I’m not 100% sure) since I have this return message:
`error_no 0 detail_no 0 user_data 0x0000000000000000^`

but for my function __LoadPngImageFromFile_ I don’t know how to use it...

Thanks 
Nicolas

**Comment by apnadkarni** [comment]
This would require two things in libcffi - support for callback functions, which I am afraid is not implemented in the current release and will not be implemented until v1.1. Sorry. The second thing is the use of longjmp which is inherently non-portable. Can you tell me what library you are looking to wrap so I can see if there is an alternative?

**Comment by nico-robert** [comment]
I'm trying to wrap this library : [libharu](http://libharu.org)  and what I’m trying to understand [Error handling](https://github.com/libharu/libharu/wiki/Error-handling) ([Help Sourceforge](http://libharu.sourceforge.net/error_handling.html))
Github : [Repo Github](https://github.com/libharu/libharu)

Regards 
Nicolas

**Comment by apnadkarni** [comment]
I had a quick look. I'll try to cook up a more complete example later this weekend if I have time. In the meanwhile, here are the basic steps.

Don't use the error handler mechanism. That is really more suitable for C++ and languages with native exception handling. The setjmp/longjmp mechanism is C is generally not a good idea. Instead do the explicit error handling as described in the link you posted. From C this would look something like

```
HPDF_Image pdfImage = HPDF_LoadPngImageFromFile (pdf, const char *filename);
if (pdfImage == NULL) {
   /* Error handling */
  HPDF_STATUS error_no = HPDF_GetError(pdf);
  HPDF_ResetError(pdf); /* Important - see your link */
 ... do something with the error code...
} else {
  ...Success...
}
```

From Tcl/CFFI, translating the above should be straightforward. You can either do a direct translation of the above code assuming the library object is `haru` ("pseudocode" not actually tested):

```
cffi::alias define HPDF_STATUS uint  ;# You have to see what the actual HPDF_STATUS type is
haru function HPDF_ResetError void {pdf pointer.HPDF_doc}
haru function HPDF_GetError HPDF_STATUS {pdf pointer.HPDF_doc}
haru function HPDF_LoadPngImageFromFile pointer.HPDF_Image {pdf pointer.HPDF_doc filepath string}

set image [HPDF_LoadPngImageFromFile $pdf $path]
if {[cffi::isnull $image]} {
  set error [HPDF_GetError $pdf]
  HPDF_ResetError $pdf
  error "Got error $error loading $path"
} else {
  ...success...
}
```

Better still, instead of having to check errors explicitly on each call, set up CFFI `onerror` handlers. Something like 

```
cffi::alias define HPDF_STATUS uint  ;# You have to see what the actual HPDF_STATUS type is
haru function HPDF_ResetError void {pdf pointer.HPDF_doc}
haru function HPDF_GetError HPDF_STATUS {pdf pointer.HPDF_doc}
haru function HPDF_LoadPngImageFromFile {pointer.HPDF_Image nonzero {onerror ErrorHandler}} {pdf pointer.HPDF_doc filepath string}
proc ErrorHandler {func_result inparams outparams} {
  set pdf [dict get $inparams pdf]
  set error [HPDF_GetError $pdf]
  error "Got error $error loading $path"
}

## Now with the above definitions, you do not need to explicitly check errors
set image [HPDF_LoadPngImageFromFile $pdf $path]
...success else an exception would have been raised on above line...
```
For more on error and exception handling, see the following:

- Basics: https://cffi.magicsplat.com/#Error%20checking
- Reference: https://cffi.magicsplat.com/cffi_Concepts.html#::Concepts-Error%20handling
- A complete example: https://github.com/apnadkarni/tcl-cffi/blob/d1c3310a3802fc85ea2237882852df53de6ab125/examples/libzip.tcl#L172

**Comment by nico-robert** [comment]
Thank you for this additional explanation but `tclsh` crash with the same message... `Bus error: 10`
```
Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   ???                           	0x00007fa17b01ae40 0 + 140331530169920
1   libhpdf-2.3.0.dylib           	0x000000010626fb4d HPDF_LoadPngImageFromFile + 103
2   libcffi1.0a8.dylib            	0x0000000106259455 dcCall_x64_sysv + 97
3   ???                           	0x00007ffee9b902d0 0 + 140732819636944
```
What I am sure is that if my path is not correct, the following functions are not read since `tclsh` crashes. Like your example `cffi::pointer isnull $image` this command is not even read.
The error may come from me which incorrectly defines my returns of functions...  I don't know

Regards
Nicolas

**Comment by apnadkarni** [comment]
@nico-robert if you can share your code where you are defining the cffi wrappers, I can take a look.

**Comment by nico-robert** [comment]
I can...
```
proc ErrorHandler {func_result inparams outparams} {
  set pdf [dict get $inparams pdf]
  set error [HPDF_GetError $pdf]
  error "Got error $error loading $path"
}
package require cffi

set imagedir [file dirname [info script]] ; # image demo dir

cffi::dyncall::Library create HPDF libhpdf-2.3.0.dylib
cffi::alias load C

cffi::Struct create error_handler {
    error_no  ulong
    detail_no ulong
    user_data {pointer unsafe}
}

cffi::alias define HPDF_ErrorHandler {struct.error_handler byref}

cffi::alias define HPDF_HANDLE int
cffi::alias define HPDF_STATUS ulong
cffi::alias define HPDF_Image  HPDF_HANDLE

HPDF stdcalls {

    HPDF_New                  pointer.HPDF_HANDLE                                 {user_error_fn {HPDF_ErrorHandler out} user_data int}
    HPDF_Free                 {pointer.HPDF_HANDLE nonzero}                       {pdf pointer.HPDF_HANDLE}
    HPDF_GetError             pointer.HPDF_STATUS                                 {pdf pointer.HPDF_HANDLE}
    HPDF_GetErrorDetail       pointer.HPDF_STATUS                                 {pdf pointer.HPDF_HANDLE}
    HPDF_ResetError           void                                                {pdf pointer.HPDF_HANDLE}
    HPDF_LoadPngImageFromFile {pointer.HPDF_Image nonzero {onerror ErrorHandler}} {pdf pointer.HPDF_HANDLE filename string}

}

set pdf   [HPDF_New err 0]
set image [HPDF_LoadPngImageFromFile $pdf [file join $imagedir pngsuite "basn3p02.png"]]
```
attached the image folder
[pngsuite.zip](https://github.com/apnadkarni/tcl-cffi/files/7575058/pngsuite.zip)

Thanks
Nicolas

**Comment by apnadkarni** [comment]
Nicolas, here is a [working example ](https://github.com/apnadkarni/tcl-cffi/blob/main/examples/haru.tcl).

I'll comment on what was wrong with your code later.

**Comment by nico-robert** [comment]
Your example works, thank you... I look forward to comments on my code 
Have a good day.

Regards
Nicolas

**Comment by apnadkarni** [comment]
Some comments on your code.

First, you do not need to define any error handler callbacks. For now, do explicit checks as shown in my sample. Later still, I'll rewrite my sample with using exceptions at the Tcl level.

With regard to the aliases ...

> cffi::alias define HPDF_HANDLE int

If you look at the hpdf.h definitions you will see that `HPDF_HANDLE` is a `void *`, not an int. Defining and passing it as an `int` will lead to memory corruption and crashes sooner or later. So the alias should really be

```
cffi::alias define HPDF_HANDLE pointer
```

Then, the following alias

> cffi::alias define HPDF_Image  HPDF_HANDLE

is not wrong but better written as  

> cffi::alias define HPDF_Image pointer.HPDF_Image

This provides type safety so you cannot pass another type (like `HPDF_Page`) as an argument to a function expecting a `HPDF_Image`. If you define all as `HPDF_HANDLE` that allows them to be interchangeable which is dangerous. In fact, I would suggest not defining `HPDF_HANDLE` at all but instead define type-specific aliases like `HPDF_Doc`, `HPDF_Page`, `HPDF_Image` as `pointer.HPDF_Doc` etc.

Once you have defined the alias as above, use directly as `HPDF_Image`, not `pointer.HPDF_Image` since the alias is already
typed as a pointer.

The function HPDF_Free frees up the pointer to the `HPDF_Doc` and should not be accessed again.
>     HPDF_Free                 {pointer.HPDF_HANDLE nonzero}                       {pdf pointer.HPDF_HANDLE}

So it is better to add the `dispose` annotation to the argument so CFFI knows that pointer should not be accessed again. This safeguards against user-after-free errors. Also, the function returns `HPDF_STATUS`, not a pointer, so the definition should read something like

```
HPDF_Free HPDF_STATUS {hpdf pointer.HPDF_Doc}
```

Moreover, all functions returning `HPDF_STATUS`, for example
>     HPDF_GetError             pointer.HPDF_STATUS                                 {pdf pointer.HPDF_HANDLE}

should have that as the return type, not `pointer.HPDF_STATUS`. They are not returning pointers, they are returning the value itself.

/Ashok

**Comment by apnadkarni** [comment]
By the way, if your interest is in generating PDF, you may consider the pure-Tcl pdf4tcl package instead of having to write your own using cffi. See the [repository](https://sourceforge.net/projects/pdf4tcl/) and the [docs](http://pdf4tcl.sourceforge.net/)

**Comment by nico-robert** [comment]
Thank you  @apnadkarni for your time and your comments, I appreciate that.

_"By the way, if your interest is in generating PDF, you may consider the pure-Tcl pdf4tcl package instead of having to write your own using cffi. See the repository and the docs"_

Yes I know , but I want to embed this kind of file [U3D](https://en.wikipedia.org/wiki/Universal_3D) in my PDF and I believe that pdf4tcl is not able to do it.

**Comment by apnadkarni** [comment]
Ah, I see. I may look into haru further as I'm interested in real world libraries to drive development in CFFI and haru looks like something useful to have.

**Comment by nico-robert** [comment]
I would also like to be able to share this project on Github

**Comment by apnadkarni** [comment]
I have updated my haru example with more comments and using exceptions for errors so you do not have to check status codes on every call.

Note this needs the current repository version, not the released 1.0b0, and has some incompatibilities with previous releases. In particular, NULL pointers will now raise exceptions by default unless the `nullok` annotation is included.

/Ashok

**Comment by apnadkarni** [comment]
BTW, if you start a project on github, please let me know.

**Comment by nico-robert** [comment]
> BTW, if you start a project on github, please let me know.

I’ll be sure to let you know...
Regards
Nicolas

**Comment by nico-robert** [comment]
Hello @apnadkarni ,

Is this writing correct ?

```
# libcffi1.0b1
cffi::alias define HPDF_Outline pointer.HPDF_Outline

HPDF_CreateOutline HPDF_Outline {
    pdf     HPDF_Doc
    parent  {HPDF_Outline unsafe nullok}
    title   string
    encoded {HPDF_Encoder unsafe nullok}
}
set root [HPDF_CreateOutline $pdf NULL "Encoding list" NULL]
```
I would like to write my `cffi::alias` like this :
```
cffi::alias define HPDF_Outline {pointer.HPDF_Outline unsafe nullok}

HPDF_CreateOutline HPDF_Outline {
    pdf     HPDF_Doc
    parent  HPDF_Outline
    title   string
    encoded {HPDF_Encoder unsafe nullok}
}
```
But I don’t know if it works in a return function like `HPDF_CreateOutline` above
Since this function should not return a null pointer and instead return me an error if yes.

Regards 
Nicolas

**Comment by apnadkarni** [comment]
The above will not work for the reason you said - it will not raise an exception if a NULL pointer is returned.

You have two choices - define two aliases or better, define a partial alias as below

```
cffi::alias define HPDF_Outline {pointer.HPDF_Outline unsafe}
```

Then define the function as

```
HPDF_CreateOutline HPDF_Outline {
    pdf     HPDF_Doc
    parent  HPDF_Outline
    title   string
    encoded {HPDF_Encoder nullok}
}
```

**Comment by nico-robert** [comment]
Hello @apnadkarni ,
One day a problem... :wink:
I have defined this :
```
cffi::Struct create hpdfrect {
    left   float
    bottom float
    right  float
    top    float
}

cffi::alias define HPDF_Annotation  pointer.HPDF_Annotation
cffi::alias define HPDF_Rect {struct.hpdfrect byref}

HPDF_Page_CreateLinkAnnot HPDF_Annotation {
    page HPDF_Page
    rect HPDF_Rect
    dst  HPDF_Destination
}
```
And my code : 
```
set pdf [HPDF_New]
set index_page [HPDF_AddPage $pdf]
# my structure 
set p {left 16.0 bottom 176.0 right 24.0 top 190.0}
# my destination 
set page(0) [HPDF_AddPage $pdf]
set dst [HPDF_Page_CreateDestination $page(0)]
# and my link annotation 
set annot [HPDF_Page_CreateLinkAnnot $index_page p $dst]
```
Passing of structs by value is not supported, for this, I added `byref` in `HPDF_Rect` but when I call `HPDF_Page_CreateLinkAnnot` I have this message :
_missing value to go with key_

Again I am not sure about myself and if that is how I should proceed...

Regards
Nicolas

**Comment by apnadkarni** [comment]
Unfortunately passing structs by value is not supported because the underlying dyncall library does not support it.

Let me check with the developers about workarounds. Is there a specific platform (OS and architecture) you care particularly about?

**Comment by nico-robert** [comment]
Hello @apnadkarni ,
I would say on all platforms... but I guess if you ask me it’s not possible , currently I’m on Mac OS Big Sur 64 and Windows 10 64.

Regards 
Nicolas

**Comment by apnadkarni** [comment]
Given the uncertain time frame when dyncall upstream will have support for passing structs by value, I've started adding support for libffi as a back end which does support that functionality. Hopefully, I should be done in a couple of weeks. If you still have interest in libharu, I would suggest waiting on that particular call and proceeding with the rest for now.

/Ashok

**Comment by nico-robert** [comment]
Ok , I can wait 
Thanks

Regards
Nicolas

**Comment by apnadkarni** [comment]
I have implemented support for using the libffi library instead of dyncall. This supports passing structs by value. Please download v1.0b3 or build from the repository. Note you need libffi installed. See BUILD.md for instructions.

I am closing this ticket as it really refers to different issues. Please open a separate ticket if you have issues with passing structs by value. For questions, please use the [Discussions ](https://github.com/apnadkarni/tcl-cffi/discussions)tab.

**Comment by nico-robert** [comment]
I will try 
Thanks Ashok
